test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,508 @@
+	// Make sure bootstrap data is available and populated.
+	if machinePool.Spec.Template.Spec.Bootstrap.DataSecretName == nil {
+		log.Info("Waiting for the Bootstrap provider controller to set bootstrap data")
+		return ctrl.Result{}, nil
+	}
+
+	if machinePool.Spec.Replicas == nil {
+		machinePool.Spec.Replicas = pointer.Int32Ptr(1)
+	}
+
+	pool, err := newNodePool(r, cluster, machinePool, dockerMachinePool, log)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to build new node pool")
+	}
+
+	// if we don't have enough nodes matching spec, build them
+	if err := pool.ReconcileMachines(ctx); err != nil {
+		if errors.Is(err, &TransientError{}) {
+			log.V(4).Info("requeue in 5 seconds due docker machine reconcile transient error")
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		return ctrl.Result{}, errors.Wrap(err, "failed to reconcile machines")
+	}
+
+	if err := pool.DeleteExtraMachines(ctx); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to delete overprovisioned or out of spec machines")
+	}
+
+	dockerMachinePool.Spec.ProviderIDList = []string{}
+	for _, instance := range dockerMachinePool.Status.Instances {
+		if instance.ProviderID != nil && instance.Ready {
+			dockerMachinePool.Spec.ProviderIDList = append(dockerMachinePool.Spec.ProviderIDList, *instance.ProviderID)
+		}
+	}
+
+	dockerMachinePool.Status.Replicas = int32(len(dockerMachinePool.Spec.ProviderIDList))
+
+	if dockerMachinePool.Spec.ProviderID == "" {
+		dockerMachinePool.Spec.ProviderID = getDockerMachinePoolProviderID(cluster.Name, dockerMachinePool.Name)
+	}
+
+	dockerMachinePool.Status.Ready = len(dockerMachinePool.Spec.ProviderIDList) == int(*machinePool.Spec.Replicas)
+	return ctrl.Result{}, nil
+}
+
+type nodePool struct {
+	client            client.Client
+	cluster           *clusterv1.Cluster
+	machinePool       *clusterv1exp.MachinePool
+	dockerMachinePool *infrav1exp.DockerMachinePool
+	machines          []*docker.Machine
+	logger            logr.Logger
+}
+
+func newNodePool(kClient client.Client, cluster *clusterv1.Cluster, mp *clusterv1exp.MachinePool, dmp *infrav1exp.DockerMachinePool, log logr.Logger) (*nodePool, error) {
+	np := &nodePool{
+		client:            kClient,
+		cluster:           cluster,
+		machinePool:       mp,
+		dockerMachinePool: dmp,
+		logger:            log.WithValues("node-pool", dmp.Name),
+	}
+
+	if err := np.Refresh(); err != nil {
+		return np, errors.Wrapf(err, "failed to refresh the node pool")
+	}
+	return np, nil
+}
+
+func (np *nodePool) Refresh() error {
+	machinePoolLabels := map[string]string{"dmp": np.dockerMachinePool.Name}
+	machines, err := docker.ListMachinesByCluster(np.cluster.Name, machinePoolLabels, np.logger)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
+	}
+
+	np.machines = machines
+	return nil
+}
+
+func (np *nodePool) MachinesMatchingSpec() []*docker.Machine {
```

Comment:
nit make private?

Reply:
Also, for my understanding, in this case we are concerned about the infrastructure spec (not the kubeadmconfig part).
If yes, what about 
```suggestion
func (np *nodePool) machinesMatchingInfrastructureSpec() []*docker.Machine {
```

Reply:
renamed